core/validation/failures.py

```diff
@@ -147,10 +148,41 @@ def __str__(self):
             f'type "{self.expected_type}", got type "{self.actual_type}"'
         )
 
-    def to_user_centered_components(self) -> tuple[str, str, str, str]:
-        # Numbers - 1. Outcomes/Outputs is just numerical (e.g not m^2), 2. Funding/PSI is Monetary
-        # Dates - Programme Progress sheet: Start, Completion and Date of Most Important...
-        return "Unimplemented", "Unimplemented", "Unimplemented", "Unimplemented"
+    def to_user_centered_components(self) -> tuple[str, str, str]:
+        sheet = INTERNAL_TABLE_TO_FORM_TAB[self.sheet]
+        column, section = INTERNAL_COLUMN_TO_FORM_COLUMN_AND_SECTION[self.column]
+        expected_type = INTERNAL_TYPE_TO_USER_CENTERED_TYPE[self.expected_type]
+        actual_type = INTERNAL_TYPE_TO_USER_CENTERED_TYPE[self.actual_type]
+        if sheet == "Outcomes":
+            column, section = "F-O", "Project Outputs"
+
+        if self.expected_type == "datetime64[ns]":
+            message = (
+                f"For column {column} you entered {actual_type} when we expected {expected_type}. "
+                f"You must enter the start date in the correct format, for example, Dec-22, Jun-23"
+            )
+
+        if sheet == "PSI":
+            message = (
+                f"For column {column} you entered {actual_type} when we expected {expected_type}. "
+                f"You must enter the required data in the correct format, for example, £5,588.13 or £238,"
+                f"062.50"
+            )
+
+        if sheet == "Funding Profiles":
+            message = (
+                f"Between columns {column} you entered {actual_type} when we expected {expected_type}. "
+                f"You must enter the required data in the correct format, for example, £5,588.13 or £238,"
+                f"062.50"
+            )
+
+        if sheet in ["Project Outputs", "Outcomes"]:
+            message = (
+                f"Between columns {column} you entered {actual_type} when we expected {expected_type}. "
+                f"You must enter data using the correct format, for example, 9 rather than 9m2. Only use numbers"
+            )
```

Comment:
My original comment was:

> Do we know what this case is or is it just an unknown path? If it's the latter I don't think we should let happen silently because it may lead to us returning potentially misleading messages.

I think using if and elif still makes sense. Just for the else we should raise an Exception. One option is to create a new exception like `UnimplementedUCFailure` and then we'll know if that occurs then we haven't accounted for all possibilities

Reply:
I think that is a good Idea, might be able to reuse UnimplementedUCFailure in other cases as well. Will we log user failures from the tool so that we can get this information?

Reply:
So just to clarify I mean raising a run time exception rather than returning a different failure to the user.. If we raise an uncaught exception we'll return 500 to the frontend and have to handle that (i created a ticket for this this morning)

Reply:
I need a little bit more guidance here, currently getting errors arround circular imports when trying to raise UnimplementedUCException inside a ValidationError. Also not sure if we've nailed down the behaviour here. whether we still want to return an accordian of validation errors, or raise to a higher level error telling users to contact support

Reply:
I've pushed where I've got to with the UnimplementedUCException, I've also thought about a seperate option here

```
        else:
            message = (
                f"In one of the cells on this sheet you entered {actual_type} when we expected {expected_type}."
                f"You must enter data using the correct format, check all cells in which you entered data. If this "
                f"error persists contact support"
            )
```

giving a generic message that giving the option to contact support if the error persists. Probably more thinking required here form me.

Reply:
Yeah I think this could be an option but I'd rather make sure that we notice the scenarios we haven't implemented during QA